pkg/uroot/bb.go

```diff
@@ -25,96 +28,79 @@ import (
 )
 
 // Per-package templates.
+//
+// TODO: Generate the cmd_xx.go template using the AST, and use a function-name
+// generator to avoid collisions.
 const (
 	cmdFunc = `package main
 
 import "github.com/u-root/u-root/bbsh/cmds/{{.CmdName}}"
 
-func _forkbuiltin_{{.CmdName}}(c *Command) (err error) {
+func _forkbuiltin_{{.CmdName}}(c *Command) error {
 	{{.CmdName}}.Main()
-	return
+	return nil
 }
 
 func {{.CmdName}}Init() {
 	addForkBuiltIn("{{.CmdName}}", _forkbuiltin_{{.CmdName}})
-	{{.Init}}
+	{{.CmdName}}.Init()
 }
-`
-
-	bbsetupGo = `package {{.CmdName}}
-
-import "flag"
-
-var {{.CmdName}}flag = flag.NewFlagSet("{{.CmdName}}", flag.ExitOnError)
 `
 )
 
 // init.go
 const initGo = `package main
 
 import (
-	"flag"
-	"fmt"
 	"log"
 	"os"
 	"os/exec"
 	"path/filepath"
-	"strings"
         "syscall"
 
 	"github.com/u-root/u-root/pkg/uroot/util"
 )
 
-func usage () {
-	n := filepath.Base(os.Args[0])
-	fmt.Fprintf(os.Stderr, "Usage: %s:\n", n)
-	flag.VisitAll(func(f *flag.Flag) {
-		if !strings.HasPrefix(f.Name, n+".") {
-			return
-		}
-		fmt.Fprintf(os.Stderr, "\tFlag %s: '%s', Default %v, Value %v\n", f.Name[len(n)+1:], f.Usage, f.Value, f.DefValue)
-	})
-}
-
 func init() {
-	flag.Usage = usage
-	// This getpid adds a bit of cost to each invocation (not much really)
-	// but it allows us to merge init and sh. The 600K we save is worth it.
-	// Figure out which init to run. We must always do this.
-
-	// log.Printf("init: os is %v, initMap %v", filepath.Base(os.Args[0]), initMap)
-	// we use filepath.Base in case they type something like ./cmd
+	// Init for the command that is being run.
 	if f, ok := initMap[filepath.Base(os.Args[0])]; ok {
-		//log.Printf("run the Init function for %v: run %v", os.Args[0], f)
 		f()
 	}
 
+	// Yeah, it's weird that this stuff happens in bbsh's init.
+	// We only do it if we actually *are* init. But merging this into bbsh
+	// should happen some other way. This should just be in bbsh.Main?
+	//
+	// Maybe init should just be a normal bb command, just like the others,
+	// except with a special symlink to /init?
```

Comment:
but you want init and rush as one command. That's the issue. Two commands and you blow the rom space budget. The space problem is the source of many sins.

Reply:
You do want it as one command. But why not just compile init into rush just like all the other bb commands? Everything already gets compiled into rush, no need to make init special except for the symlink.

Reply:
Agreed. I'm actually thinking that we may want two different inits. One for source mode, one for bb mode.

```
cmds/bb/init
cmds/source/init
```

Reply:
On the other hand, if we want to allow multi-mode initramfs, maybe one init should be suitable.

Reply:
Another possibility would be a single init with a custom buildtag gating feature we don't want in bb. For example, "bgbuild.go" which we definitely don't want in bb mode would have the line `// +build !bb`. I believe Go lets you use your own buildtags with the `ctxt.BuildTags` slice or via the `-tags` option in "go build".

Reply:
That's a nice idea. Like I said, another PR.